File: hw6/matlab/hw6.m
%Halime Özge KABAK
%180403001
% Image Processing HW6
%Add Noise
%--------------------------------------------------------------------------
%Salt and Pepper Noise
% f=imread('lenna.png');
% figure
% imshow(f)
% [M, N]= size(f);
% R = saltpepper(M, N, 0.05, 0.05);
% c = find (R==0);
% gp = f;
% gp(c) = 0;
% d = find (R==1);
% gp(d) = 255;
% figure
% imshow(gp)
%For eliminating noise ==> median filter (min filter for pepper) (max filter for salt)
% %--------------------------------------------------------------------------
% % %Gaussian Noise
% S = 20; %standart deviation
% I = imread('cameraman.jpg');
% [M, N]= size(I);
% Gframe=double(I);
% Gframe = Gframe - min(Gframe(:));
% Gframe = Gframe / max(Gframe(:));
% figure;
% imshow((I))
% % J = double(I) + S.*randn(size(I));
% % figure;
% % imshow(J./255)
% n=0.1*noises('gaussian', M,N);
% J=n+Gframe;
% figure;
% imshow((J))
% % For eliminating noise ==>image averaging/ Arithmetic/Geometric/ Harmonic Mean Filter
% Kr=3;
% Kc=3;
% J=im2double(J);
% %J=(Kr*Kc)./imfilter(1./(J+eps),ones(Kr,Kc),"replicate"); %Harmonic filter
% %J=exp(imfilter(log(J),ones(Kc,Kr),"replicate")).^(1/(Kr*Kc)); %Geometric Mean Filter
% %J=wiener2(J,[3,3]); %Adaptive Filter
% figure;
% imshow((J))
% %--------------------------------------------------------------------------
% %Erlang Noise
% n=0.1*noises('erlang', M,N);
% Gframe=imread('cameraman.jpg');
% Gframe=double(Gframe);
% Gframe = Gframe - min(Gframe(:));
% Gframe = Gframe / max(Gframe(:));
% J = Gframe + n;
% figure
% imshow(J)
% %--------------------------------------------------------------------------
% %Uniform Noise
% n2=0.3*noises('uniform', M,N);
% J = Gframe + n2;
% figure
% imshow(J)
% %--------------------------------------------------------------------------
% %Rayleigh Noise
% n3=0.3*noises('rayleigh', M,N);
% J = Gframe + n3;
% figure
% imshow(J)
% %--------------------------------------------------------------------------
% %Exponential Noise
% n4=0.2*noises('exponential', M,N);
% J = Gframe + n4;
% figure
% imshow(J)
% %--------------------------------------------------------------------------
% %Lognormal Noise
% n5=0.3*noises('lognormal', M,N);
% J = Gframe + n5;
% figure
% imshow(J)
%--------------------------------------------------------------------------
%Periodic Noise
 t = imread ('cameraman.jpg');
s = size (t); % It assigns the size of t to s.
[x, y] = meshgrid (1: s (1), 1: s (2)); % x, y dimensional matrix.
p = sin (x / 3 + y / 5) +1; % sins
imshow (p)
k = (im2double (t) + p / 2) / 2;
imshow (k)
[x, y] = meshgrid (1: 256,1: 256);
p = 1 + sin (x + y / 1.5);
tp = (double (k) / 128 + p) / 8;
imshow (tp);
af = fftshift (fft2 (tp));
imshow (mat2gray (log (1 + abs (af))));
tp(88,:)=0;
tp(171,:)=0;
tp(:,156)=0;
tp(:,102)=0;
af = fftshift (fft2 (tp));
imshow (mat2gray (log (1 + abs (af))));
s=ifftshift(ifft2(af));
imshow(mat2gray((1+abs(s))))
%--------------------------------------------------------------------------
%SNR
img=f;
img=double(img(:));
ima=max(img(:));
imi=min(img(:));
ims=std(img(:));
snr=10*log((ima-imi)./ims);
disp(snr)

File: hw6/python/HW6.py
import cv2
from noises import noises, salt_papper, add_periodic
import numpy as np
import matplotlib.pyplot as plt
img = cv2.imread('cameraman.jpg')
img = cv2.cvtColor(img,cv2.COLOR_BGR2GRAY)
cv2.imshow('original image',img)
Gframe=np.double(img)
Gframe = Gframe - Gframe.min()
Gframe = Gframe / Gframe.max()
[M,N]=img.shape
#R=0.3*noises('uniform', M,N,0,1)
#R=0.1*noises('gaussian', M,N,0,1)
#R=0.3*noises('lognormal', M,N,1,0.25)
#R=0.3*noises('rayleigh', M,N,0,1)
#R=0.2*noises('exponential', M,N,1,1)
#R=0.1*noises('erlang', M,N,2,5)
#R=R+Gframe
#R=salt_papper(img,0.05)










periodic = add_periodic(img).astype(int)
plt.figure(figsize=(6,6))
plt.imshow(periodic,cmap="gray")
plt.title("Periodic Noise")
plt.show()
from scipy import fftpack
import numpy.fft as fp
w = 10
h = 10

im = periodic
F1 = fftpack.fft2((im).astype(float))
F2 = fftpack.fftshift(F1)
for i in range(60, w, 135):
    for j in range(100, h, 200):
        if not (i == 330 and j == 500):
            F2[i-10:i+10, j-10:j+10] = 0
    for i in range(0, w, 135):
        for j in range(200, h, 200):
            if not (i == 330 and j == 500):
                F2[max(0,i-15):min(w,i+15), max(0,j-15):min(h,j+15)] = 0

plt.figure(figsize=(6,6))
plt.title("Spectrum")
plt.imshow( (20*np.log10( 0.1 + F2)).astype(int), cmap=plt.cm.gray)
plt.show()
F1[88,:]=0
F1[171,:]=0
F1[:,156]=0
F1[:,102]=0
F2 = fftpack.fftshift(F1)
im1 = fp.ifft2(fftpack.ifftshift(F2)).real
plt.figure(figsize=(6,6))
plt.title("Recovered Image")
plt.imshow(im1, cmap='gray')
plt.show()













# waitKey() waits for a key press to close the window and 0 specifies indefinite loop
cv2.waitKey(0)

# cv2.destroyAllWindows() simply destroys all the windows we created.
cv2.destroyAllWindows()
